### PR TITLE
Updates hydroelastic field computation for the ellipsoid and meshes.

### PR DESCRIPTION
**Pull request by amcastro-tri**
Additional experimentation and testing revealed better ways to compute the pressure fields for these gometries.

**For the elliposoid field**, a convex geometry, the inititial method to compute pressure field would become singular if the margin was equal or larger than the minimum semi-axis lenght. This is obviously wrong. For convex geometries inflation and pressure field defintion must be valid for any value of the margin. 
The new filed then pairs up in #21773 with an inflated ellipsoid of semi-axis aᵢ+δ so that the zero-level pressure field exactly matches the original geometry.

**For the mesh field**, being non-convex, the inflation process might lead to cracks that, though negligible for the computation of forces, introduce topological gaps that completely change the computation of the distance function used to define the pressure field. We resolve this in #21773 by defining the pressure field on the original mesh (with good topology) and "warping" it to the inflated mesh. Thereofre this requires the new field computatioin to reason about this warping with an approximation of the distance field in the inflate mesh from the distance field on the non-inflated mesh.

A mouthful really, but the changes are quite simple to implement. Testing in Anzu even shows that we can even deal with very thin geometry, which is even far beyond any of our original expectations.

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21777)

### CODE REVIEW

**Review by amcastro-tri — COMMENTED** [review]
+@SeanCurtis-TRI for feature review please. Part of # #21773. I figured we could make quick progress on this and clean up #21773 for later review.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-:-O3Z4z5Y0hugiZqC7hRg:buv2tzg)* status: LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers, missing label for release notes (waiting on @amcastro-tri)

**Review by amcastro-tri — COMMENTED** [review]
-@SeanCurtis-TRI, since Joe will assign another bigger PR to you.

+@joemasterjohn for feature review please.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-:-O3Z7J0w31lD0ZNT4tRo:b-d0v3gm)* status: LGTM missing from assignee joemasterjohn, needs platform reviewer assigned, needs at least two assigned reviewers

**Review by joemasterjohn — COMMENTED** [review]
:lgtm: with a few requests.

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-:-O3cwodt1IHYZypo95T3:b-334hhk)* status: 9 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers

___
*[`geometry/proximity/make_ellipsoid_field.h` line 84 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-O3cDS6QBJEcc5cw1SWA:-O3cDS6QBJEcc5cw1SWB:b-k7uq7c) ([raw file](https://github.com/RobotLocomotion/drake/blob/d2c33757febb5a179da1d03914b3a4617bae9a28/geometry/proximity/make_ellipsoid_field.h#L84)):*
> ```objc
>     // Scale V in the ellipsoid to U in the unit sphere.
>     const Vector3<T> r_EU = scale.cwiseProduct(r_EV);
>     T extent = T(1.0) - r_EU.norm();  // Zero margin extent.
> ```

Comment needs update, no longer the "zero-margin extent".



_Code quote:_
```objc
// Zero margin extent.
```
___
*[`geometry/proximity/make_ellipsoid_field.h` line 87 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-O3cDjVU8aPnoIJwTQRy:-O3cDjVU8aPnoIJwTQRz:b6oavn3) ([raw file](https://github.com/RobotLocomotion/drake/blob/d2c33757febb5a179da1d03914b3a4617bae9a28/geometry/proximity/make_ellipsoid_field.h#L87)):*
> ```objc
>     if (abs(extent) < kExtentEpsilon) {
>       extent = 0.0;
>     }
> ```

In the comment above the declaration of `kExtentEpsilon` you mention that you only clamp near zero extent to zero in the margin  = 0 case, but that is not implemented here. 



_Suggestion:_
```objc
    if (margin == 0 && abs(extent) < kExtentEpsilon) {
      extent = 0.0;
    }
```
___
*[`geometry/proximity/make_mesh_field.cc` line 77 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-O3cm5He-OtrRu35uja-:-O3cm5He-OtrRu35uja0:b54nkz3) ([raw file](https://github.com/RobotLocomotion/drake/blob/d2c33757febb5a179da1d03914b3a4617bae9a28/geometry/proximity/make_mesh_field.cc#L77)):*
> ```cpp
>     // inflated surface as the distance to the original surface plus the margin.
>     // Given the algorithm we use to inflate meshes, this is a lower bound on
>     // the correct distance. This correction only applies to interior vertices.
> ```

nit: We reasoned that this was a lower bound when working this out on paper, but never came up with a succinct proof.
Do you have a short way to show that this is indeed true?



_Code quote:_
```cpp
    // Given the algorithm we use to inflate meshes, this is a lower bound on
    // the correct distance. This correction only applies to interior vertices.
```
___
*[`geometry/proximity/test/make_ellipsoid_field_test.cc` line 99 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-O3csxFQ0-N97HGNZwx6:-O3csxFQ0-N97HGNZwx7:b-3gq3u9) ([raw file](https://github.com/RobotLocomotion/drake/blob/d2c33757febb5a179da1d03914b3a4617bae9a28/geometry/proximity/test/make_ellipsoid_field_test.cc#L99)):*
> ```cpp
>   // not exactly on the surface of the ellipsoid due to numerical roundings.
>   // We do not want to use the coarsest mesh (octahedron) since all vertices
>   // are exactly on the coordinate axes.
> ```

nit: Seems like you're only testing the vertices exactly on the axes now. So either this comment needs update, or just test the octahedron mesh?



_Code quote:_
```cpp
  // We do not want to use the coarsest mesh (octahedron) since all vertices
  // are exactly on the coordinate axes.
```
___
*[`geometry/proximity/test/make_ellipsoid_field_test.cc` line 110 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-O3cvb0O-I_4MJkvk9OF:-O3cvb0O-I_4MJkvk9OG:b-5hp7ll) ([raw file](https://github.com/RobotLocomotion/drake/blob/d2c33757febb5a179da1d03914b3a4617bae9a28/geometry/proximity/test/make_ellipsoid_field_test.cc#L110)):*
> ```cpp
> 
>   // If `p` is a point along one of the axes, it returns the index of the
>   // non-zero coordinate. Otherwise it returns a negative value.
> ```

nit: Slightly incorrect, this detects points that are coincident with "vertices/co-vertices" of the ellipse, so just boundary vertices of our mesh that are on an axis. Interior points that lie on the axes do not return true.



_Code quote:_
```cpp
  // If `p` is a point along one of the axes, it returns the index of the
  // non-zero coordinate. Otherwise it returns a negative value.
```
___
*[`geometry/proximity/test/make_ellipsoid_field_test.cc` line 111 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-O3cuI-9EhC7PQjYCVrN:-O3cuI-9EhC7PQjYCVrO:b-n77p1b) ([raw file](https://github.com/RobotLocomotion/drake/blob/d2c33757febb5a179da1d03914b3a4617bae9a28/geometry/proximity/test/make_ellipsoid_field_test.cc#L111)):*
> ```cpp
>   // If `p` is a point along one of the axes, it returns the index of the
>   // non-zero coordinate. Otherwise it returns a negative value.
>   auto point_on_axis = [](const Vector3d sizes, const Vector3d& p) {
> ```

nit: Name is a bit misleading. It's not just testing that the point is on the axis, but that it lies exactly at the "vertex" or one of the "co-vertices" of the ellipsoid.



_Code quote:_
```cpp
  auto point_on_axis = [](const Vector3d sizes, const Vector3d& p) {
```
___
*[`geometry/proximity/test/make_ellipsoid_field_test.cc` line 135 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-O3ctsJG7lLnO-E7vu9c:-O3ctsJG7lLnO-E7vu9d:brn7iqg) ([raw file](https://github.com/RobotLocomotion/drake/blob/d2c33757febb5a179da1d03914b3a4617bae9a28/geometry/proximity/test/make_ellipsoid_field_test.cc#L135)):*
> ```cpp
>     }
>   }
>   EXPECT_EQ(num_points_tested, 6);
> ```

Given `point_on_axis` expects bitwise equivalence in the point coordinates, you're relying on some internal details of the ellipsoid mesh generation code (e.g. that it places 6 vertices on the axes and subdivides, etc). That assumption should at least be documented here.



_Code quote:_
```cpp
EXPECT_EQ(num_points_tested, 6);
```
___
*[`geometry/proximity/test/make_mesh_field_test.cc` line 58 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-O3cwDamC8wUe-uJwi0-:-O3cwDamC8wUe-uJwi00:bxs3jvi) ([raw file](https://github.com/RobotLocomotion/drake/blob/d2c33757febb5a179da1d03914b3a4617bae9a28/geometry/proximity/test/make_mesh_field_test.cc#L58)):*
> ```cpp
> 
>   // MakeVolumeMeshPressureField() is used to generate fields with margin but,
>   // unless all other convex geometries, it is meant to work on the original
> ```

nit: typo



_Suggestion:_
```cpp
unlike all other convex geometries,
```
___
*[`geometry/proximity/test/make_mesh_field_test.cc` line 61 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-O3cwOz91MGLf9fdMMcn:-O3cwOz91MGLf9fdMMco:b-2ogir6) ([raw file](https://github.com/RobotLocomotion/drake/blob/d2c33757febb5a179da1d03914b3a4617bae9a28/geometry/proximity/test/make_mesh_field_test.cc#L61)):*
> ```cpp
>   // non-inflated mesh. It computes the field by approximating distances d̃ in
>   // the inflated mesh from distances d in the non-inflated mesh as d̃ = d + δ,
>   // with δ the margin. Therefore if H is the elastic foundation depth of the
> ```

nit: worth mentioning "except for boundary vertices which are explicitly set to `d̃ = 0`"



_Code quote:_
```cpp
  // non-inflated mesh. It computes the field by approximating distances d̃ in
  // the inflated mesh from distances d in the non-inflated mesh as d̃ = d + δ,
  // with δ the margin. Therefore if H is the elastic foundation depth of the
```

**Review by joemasterjohn — COMMENTED** [review]
+@ggould-tri for platform review, please.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-:-O3cwvY1AYEHxc79dmsU:bcav8r0)* status: 9 unresolved discussions, LGTM missing from assignee ggould-tri(platform)

**Review by ggould-tri — COMMENTED** [review]
This seems like `release notes: fix` to me since you've identified actual problems that were previously present and as of this PR no longer are.

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-:-O3d0iSIB1hMo7KFtEVt:bti3y7g)* status: 9 unresolved discussions, LGTM missing from assignee ggould-tri(platform)

**Review by amcastro-tri — COMMENTED** [review]
Question @ggould-tri, I didn't flag it as `fix` because the new feature is not yet wired up to the rest of Drake. Thus even if a fix, users won't see any change yet (this gets wired in #21773). Do you still recommed flagging as `fix`?

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-:-O3dWkoO-mYFgGAkzyHp:b-itjrcz)* status: 4 unresolved discussions, LGTM missing from assignee ggould-tri(platform)

___
*[`geometry/proximity/make_ellipsoid_field.h` line 84 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-O3cDS6QBJEcc5cw1SWA:-O3dX6mL5e3_MDW25JwR:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/d2c33757febb5a179da1d03914b3a4617bae9a28/geometry/proximity/make_ellipsoid_field.h#L84)):*
<details><summary><i>Previously, joemasterjohn (Joe Masterjohn) wrote…</i></summary><blockquote>

Comment needs update, no longer the "zero-margin extent".
</blockquote></details>

Done.
___
*[`geometry/proximity/make_ellipsoid_field.h` line 87 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-O3cDjVU8aPnoIJwTQRy:-O3dXVX-Er8JBYaRiOb8:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/d2c33757febb5a179da1d03914b3a4617bae9a28/geometry/proximity/make_ellipsoid_field.h#L87)):*
<details><summary><i>Previously, joemasterjohn (Joe Masterjohn) wrote…</i></summary><blockquote>

In the comment above the declaration of `kExtentEpsilon` you mention that you only clamp near zero extent to zero in the margin  = 0 case, but that is not implemented here.
</blockquote></details>

Done.
___
*[`geometry/proximity/make_mesh_field.cc` line 77 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-O3cm5He-OtrRu35uja-:-O3dXqkb4965IwBckvbB:b-4c9xum) ([raw file](https://github.com/RobotLocomotion/drake/blob/d2c33757febb5a179da1d03914b3a4617bae9a28/geometry/proximity/make_mesh_field.cc#L77)):*
<details><summary><i>Previously, joemasterjohn (Joe Masterjohn) wrote…</i></summary><blockquote>

nit: We reasoned that this was a lower bound when working this out on paper, but never came up with a succinct proof.
Do you have a short way to show that this is indeed true?
</blockquote></details>

The reasoning was that `MakeInflatedMesh()` moves faces outwards at least distance δ. Now, I am not sure that is enough to make a general statement about this lower bound. A rigorous proof seems pretty hard to achieve. 
The intuition I believe still holds, but it might not for extreme curvatures or near degenerate cases. 

TLDR; I removed this statement.
___
*[`geometry/proximity/test/make_ellipsoid_field_test.cc` line 135 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-O3ctsJG7lLnO-E7vu9c:-O3daeQD3UJkNbySSs2t:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/d2c33757febb5a179da1d03914b3a4617bae9a28/geometry/proximity/test/make_ellipsoid_field_test.cc#L135)):*
<details><summary><i>Previously, joemasterjohn (Joe Masterjohn) wrote…</i></summary><blockquote>

Given `point_on_axis` expects bitwise equivalence in the point coordinates, you're relying on some internal details of the ellipsoid mesh generation code (e.g. that it places 6 vertices on the axes and subdivides, etc). That assumption should at least be documented here.
</blockquote></details>

Done.

**Review by joemasterjohn — COMMENTED** [review]
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21777)* status: LGTM missing from assignee ggould-tri(platform)

___
*[`geometry/proximity/make_mesh_field.cc` line 77 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-O3cm5He-OtrRu35uja-:-O3dcGUO7lNHKe6nlmRs:brxp9sq) ([raw file](https://github.com/RobotLocomotion/drake/blob/d2c33757febb5a179da1d03914b3a4617bae9a28/geometry/proximity/make_mesh_field.cc#L77)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

The reasoning was that `MakeInflatedMesh()` moves faces outwards at least distance δ. Now, I am not sure that is enough to make a general statement about this lower bound. A rigorous proof seems pretty hard to achieve. 
The intuition I believe still holds, but it might not for extreme curvatures or near degenerate cases. 

TLDR; I removed this statement.
</blockquote></details>

Sounds good.

**Review by ggould-tri — COMMENTED** [review]
Ah, I didn't appreciate that this isn't wired.  Quite right.
:lgtm:

> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21777#-:-O3h3UpaC6ckSKaKy_oc:b-fypki0)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees ggould-tri(platform),joemasterjohn
